### ISSUE TITLE
Vertex shader regression in 3.2.2

### ISSUE THREAD

**Issue by hmans** [godotengine/godot#39651]
...or, possibly, my own incompetence? I've searched existing Issues for shader-related regressions and have found quite a bunch, but none of them sounded like what I'm seeing here.

**Godot version:**
3.2.2.rc1 and .rc2

**OS/device including version:**
Linux, GLES2

**Issue description:**
I've attached a simple project with a scene containing three objects. These objects are simple red circles animated with an equally simple vertex shader that "wobbles" them over time. In the editor, this looks like this:

![image](https://user-images.githubusercontent.com/1061/85057369-f619ef80-b1a0-11ea-8105-3cd5603ddeeb.png)

In 3.2.1, the game renders correctly, with the red blobs mostly keeping their shape, and only wobbling slightly:

![image](https://user-images.githubusercontent.com/1061/85057557-34afaa00-b1a1-11ea-896c-3fc3d03915ce.png)

In 3.2.2, the wobbling is all off, with the wobble effect scaling weirdly:

![image](https://user-images.githubusercontent.com/1061/85057431-0631cf00-b1a1-11ea-89e9-cac72565dce8.png)

Also, in my main project (where these were extracted from), the objects move, and for the first second or so while they're moving, they're being rendered in the correct dimensions, and suddenly they scale up as displayed above.

**Shader code:**

For reference, here's the vertex shader code. It's not particularly clever, but it does the job. Well, in 3.2.1, it does. :b

```glsl
shader_type canvas_item;

uniform float x_wobble = 50f;
uniform float y_wobble = 25f;

void vertex() {
	VERTEX.x += (UV.x - 0.5) * sin(TIME * 5f) * x_wobble;
	VERTEX.y += (UV.y - 0.5) * cos(TIME * 5f) * y_wobble;
}
```

**Steps to reproduce:**

- Download the attached project
- Run in 3.2.1, observe that everything looks fine
- Run in 3.2.2, observe that the shaded objects are scaled incorrectly

**Minimal reproduction project:**
[shader-bug-example.zip](https://github.com/godotengine/godot/files/4800466/shader-bug-example.zip)

**Comment by clayjohn** [comment]
Can you run the project with batching turned off? You can set it in the project settings under ``rendering/batching/options/use_batching``.

**Comment by avencherus** [comment]
@clayjohn With batching they stretch quite a bit.  Without batching the transformations are much different and far less stretching.

![image](https://user-images.githubusercontent.com/13004169/85068778-a3016600-b1bb-11ea-8d49-7f195c9f5b22.png)

**Comment by clayjohn** [comment]
Looks like a missed case where batching should fail. 

@lawnjelly

**Comment by hmans** [comment]
> Can you run the project with batching turned off? You can set it in the project settings under `rendering/batching/options/use_batching`.

I can confirm that disabling batching fixes this on my system, yes.

> Looks like a missed case where batching should fail.

@clayjohn, are you saying that this is an instance where batching should not be applied? If this is what you're saying, can I ask you to elaborate? I would have -- very naively -- assumed that this (same texture, same shader) would have been an obvious use case for batching.

**Comment by clayjohn** [comment]
@hmans in order to batch objects together the position of each object is "baked" into the vertex position. However, that clearly can't be done here as you are changing vertex positions in the shader. 

That being said, instead of breaking batches we may be able to fix it another way. Looking at your code again I see that this may just be a situation where UVs are broken as a result of the batch. So breaking batches may not be necessary.

**Comment by lawnjelly** [comment]
Yes I think this is a special case because it is both reading and writing in the same expression (+=). It should be easy enough to fix.

This is confirmed because it turns out you can fix it in the shader by changing the expression to:
```
VERTEX = VERTEX + blah
```

**Comment by lawnjelly** [comment]
This has also (by chance) uncovered a really weird bug. When you run your example with batching from the editor, the bug exhibits (it batches the 3). When you run it from outside the editor, it doesn't batch the 3, and the bug doesn't exhibit!

I've tested and confirmed the batching is working for other cases outside the editor. This is more interesting than the original bug, it suggests something is behaving differently not in the editor - maybe the shader compiler, or an uninitialized variable. I am investigating! :grinning:

**Comment by akien-mga** [comment]
Fixed by #39669.

**Comment by hmans** [comment]
Thanks!
